server/src/lib/id.ts

```diff
@@ -0,0 +1,9 @@
+import { Request } from "express";
+import crypto from "crypto";
+
+export function getId(req: Request) {
+  const userAgent = req.headers["user-agent"];
+  const ip = req.headers["x-forwarded-for"] || req.connection.remoteAddress;
```

Comment:
In my experience implementing device fingerprinting, we tended to avoid using the IP address. In general, device fingerprinting is a highly complicated and non-trivial problem, however, the usage of IP address tends to generate some truly unexpected results considering how easy it is to spoof the `x-forward-for` header: [reference](https://portswigger.net/kb/issues/00400110_spoofable-client-ip-address). This is more of a FYI than anything else since for such a simple application it doesn't really matter, and we don't NEED to uniquely identify users across IPs, however, I think simply using `userAgent` instead of the combination with `ip` is actually a better identifier (although still far from perfect).